### PR TITLE
Updated Main-project_tests with new tests according to Lesson 4_Opera…

### PR DESCRIPTION
**Pull request by Yulson**
@d-konovalov, in this commit I've completed Task2 from Lesson 4 (Operators and Constructions).
2. add to your solution next test:
- create project
- make a random action: create or not a new bug issue
- open project page,
- open issues tab
- if there is a bug present then add self to watchers
- if no then create a new bug issue and then add self to watchers
- Verify that there is a bug issue and that current user is a watcher

### CODE REVIEW

File: Main_project_Helpers.rb [hunk 1]
```diff
@@ -100,18 +91,49 @@ def create_new_issue(issue_data)
...
+	def open_issues_tab(project)
+		open_project(project)
+		@wait.until {@driver.find_element(:css, '.issues').displayed?}
+		@driver.find_element(:css, '.issues').click
+	end
 
 
+	def is_issues_tab_empty?(project)
+		open_issues_tab(project)
+		if @driver.find_element(:css, '.nodata').displayed?
```
Comment (d-konovalov) [inline Main_project_Helpers.rb]: ```if``` and ```rescue``` statements can be replaced with:



```ruby
@driver.find_elements(:css,'.nodata').empty?
```

File: Main_project_Helpers.rb [hunk 2]
```diff
@@ -100,18 +91,49 @@ def create_new_issue(issue_data)
...
+		return false
+    end
+
+
+	def create_or_not_create_new_issue(project)
+		open_project(project)
+		random_action = rand(0..1)
+		if random_action == 1
+			issue = create_new_issue({})
+			open_issues_tab(project)
```
Comment (d-konovalov) [inline Main_project_Helpers.rb]: if you have this line in both ```if``` and ```else``` blocks please place it after ```if``` statement
you won't need ```else``` block at all then

File: Main_project_Helpers.rb [hunk 3]
```diff
@@ -1,17 +1,16 @@
...
-
+require_relative 'Project'
+require_relative 'Issue'
 
 
  module Main_project_Helpers
 
 
 
 	def register_user(user_data)
```
Comment (d-konovalov) [inline Main_project_Helpers.rb]: if you use this as an argument:



```
user_data=nil
```

you won't need to call this method like this:


```
register_user({})
```

just


```
register_user
```
and that's it

(this is a mistake)

File: Main_project_Tests_TestUnit.rb [hunk 4]
```diff
@@ -15,134 +15,187 @@ def setup
...
+	end
+
+
+	def test_log_in
+	   	user = register_user({})
+	   	log_out
+	   	log_in(user)
+
+	  	expected_login = user.login
+	  	actual_login = @driver.find_element(:css, '[class="user active"]').text
```
Comment (d-konovalov) [inline Main_project_Tests_TestUnit.rb]: this is not the best css locator
if this ```'.user'``` or this ```'.active'``` doesn't work you can use this one ```'.user.active'```

File: Main_project_Tests_TestUnit.rb [hunk 5]
```diff
@@ -15,134 +15,187 @@ def setup
...
+   		expected_notice = 'Successful creation.'
+   	    actual_notice = @driver.find_element(:css, '#flash_notice').text
+  	 	assert_equal(expected_notice, actual_notice)
+   	end
+
+
+	def test_create_bug
+	    user = register_user({})
+	    project = create_new_project({})
+	    issue = create_new_issue({type: '1'})
```
Comment (d-konovalov) [inline Main_project_Tests_TestUnit.rb]: please remove curly braces, we don't need them here

File: Main_project_Tests_TestUnit.rb [hunk 6]
```diff
@@ -15,134 +15,187 @@ def setup
...
-	#     assert_true(actual_array_subjects.include? expected_subject)	
-	# end
-
-	# def test_create_feature
-	# 	user = register_user({})
-	#     project = create_new_project({})
-	#     issue = create_new_issue({type: '2'})
+	    expected_subject = issue.name
+	    actual_array_subjects = []
+	    actual_array_subjects = @driver.find_elements(:css, '.subject').map {|element| element.text}
```
Comment (d-konovalov) [inline Main_project_Tests_TestUnit.rb]: this ```map``` statement also can be simplified if you use 'syntax sugar' like this


```ruby
map(&:text)
```

and btw
when you trying to find name for variable in cases like this (when its 'life' period is very short) please use shorter names. ```el``` would be enough

File: Main_project_Tests_TestUnit.rb [hunk 7]
```diff
@@ -15,134 +15,187 @@ def setup
...
+	    assert_true(actual_array_subjects.include? expected_subject)
+	end
+
+
+    def test_create_issues
+     	user = register_user({})
+	    project = create_new_project({})
+	    issue1 = create_new_issue({type: '1'})
```
Comment (d-konovalov) [inline Main_project_Tests_TestUnit.rb]: remove curly braces

File: Main_project_Tests_TestUnit.rb [hunk 8]
```diff
@@ -15,134 +15,187 @@ def setup
...
+	    issue2 = create_new_issue({type: '2'})
+	    issue3 = create_new_issue({type: '3'})
      	
- #     	actual_array_subjects = []
- #     	actual_array_subjects = @driver.find_elements(:css, '.subject').map {|element| element.text}
+     	actual_array_subjects = []
```
Comment (d-konovalov) [inline Main_project_Tests_TestUnit.rb]: this line is redundant

File: Main_project_Tests_TestUnit.rb [hunk 9]
```diff
@@ -15,134 +15,187 @@ def setup
...
-end
\ No newline at end of file
+	# Create project, make a random action: create or not a new bug issue, open project page, open issues tab, if there is a bug present then add self to watchers if no then create a new bug issue and then add self to watchers
+	# Verify that there is a bug issue and that current user is a watcher
+
+	def test_create_issue_and_add_self_to_watchers
+		user = register_user({})
+		project = create_new_project({})
+		issue = create_or_not_create_new_issue(project)
+		if is_issues_tab_empty?(project) == true
```
Comment (d-konovalov) [inline Main_project_Tests_TestUnit.rb]: ```== true``` is redundant

File: Main_project_Tests_TestUnit.rb [hunk 10]
```diff
@@ -15,134 +15,187 @@ def setup
...
+			issue = create_new_issue({})
+			add_self_to_watchers(project, issue)
```
Comment (d-konovalov) [inline Main_project_Tests_TestUnit.rb]: this line should be extracted from both ```if``` and ```else``` code blocks
and placed after ```if``` statement

File: Project.rb [hunk 11]
```diff
@@ -2,12 +2,17 @@
...
 
 	def initialize(project_data={})
 		@name = project_data[:name] || Faker::Team.name
 		@version = project_data[:version] || rand(9999)
 	end
 
-end
+	def get_url
+		directory = @name.sub(" ", "-")
+		project_url = String.new
```
Comment (d-konovalov) [inline Project.rb]: redundant

File: Project.rb [hunk 12]
```diff
@@ -2,12 +2,17 @@
...
+		project_url = "http://demo.redmine.org/projects/#{directory}"
```
Comment (d-konovalov) [inline Project.rb]: we don't need this one here```project_url =``` since we create local var we don't use anymore

**Review by d-konovalov — CHANGES_REQUESTED** [review]

File: Project.rb (hunk 12 above)
Comment (Yulson) [reply Project.rb]: @d-konovalov,  thank you a lot for a quick respond. Im proceeding with rework.

File: Main_project_Tests_TestUnit.rb [hunk 13]
```diff
@@ -50,112 +49,107 @@ def test_change_pass
...
-		project = create_new_project
-		open_project(project)
-		assert(@driver.find_element(:css, '.overview.selected').displayed?)
+    	open_project(project) || tries = 3
+
+    	begin
+    		@driver.find_element(:css, '.overview.selected').displayed?
+    	rescue ProjectNotFoundError
+    		create_new_project
+    		retry unless (tries -=1).redo
```
Comment (d-konovalov) [inline Main_project_Tests_TestUnit.rb]: I suppose ```redo``` is a kind of mistake?

**Comment by Yulson** [comment]
@d-konovalov, could you please review this pull request and finally merge it to master.
With this pull-request the following tasks should be finally done:
1) Task for Lesson 5 _"Object, Class, Method, Variables"_ (Simple Trello model with simple Unit tests)
2) Task with raising and rescuing an Exception for Lesson 7.
3) Changes in Main_project_Tests_TestUnit.rb and Main_project_Helpers.rb: changing locators, removing odd variables and methods.

File: Main_project_Helpers.rb [hunk 14]
```diff
@@ -27,91 +28,106 @@ def register_user(user_data)
...
+ 		@driver.find_element(:id, 'new_password').send_keys user.change_password
+ 		@driver.find_element(:id, 'new_password_confirmation').send_keys user.password
  		@driver.find_element(:css, '[name="commit"]').click
  	end
  	
- 	def create_new_project(project_data)
 
+ 	def create_new_project(project_data={})
  		project = Project.new unless project_data.empty?
 		project = project_data.empty? ? Project.new : Project.new(project_data)
```
Comment (d-konovalov) [inline Main_project_Helpers.rb]: you should choose which line to delete this or previous one
both of them assign a value to 'project' variable

Comment (Yulson) [reply Main_project_Helpers.rb]: @d-konovalov, I left second line. Hovewer it's longer but for me it's more clear. I also did the same changes for user and issue.

File: Main_project_Tests_TestUnit.rb [hunk 15]
```diff
@@ -5,144 +5,192 @@
...
+	end
 
 
+    def test_open_project
+    	register_user
+    	project = create_new_project
+    	open_project(project) || tries = 3
+
+    	begin
+    		@driver.find_element(:css, '.overview.selected').displayed?
```
Comment (d-konovalov) [inline Main_project_Tests_TestUnit.rb:135]: define method which will raise an Exception in specific case
then you'll be able to rescue it

**Review by d-konovalov — CHANGES_REQUESTED** [review]

**Pull request merged**